### PR TITLE
contracts: Add storage deposits

### LINKED ISSUE #9807: Deposit based storage incentivation for `pallet_contracts`
### Discussed in https://github.com/paritytech/substrate/discussions/9740

# Motivation
Up until recently `pallet_contracts` did prevent unbound state growth by charging rent or a deposit from the **contract itself**. This system was removed for various reasons: https://github.com/paritytech/substrate/pull/9669. Read the linked PR description before engaging in this discussion. We want a replacement for that system. In the following I describe one idea at a high level and then go into detail for the various areas where questions might arise.

# Overview

One rather obvious alternative to charging the contract for its own storage is to charge the caller that is responsible for creating this storage. This is distinct from and in addition to gas whose purpose it is to charge for execution time. Using ongoing payments (rent) is not really viable so it would be purely deposit based like in any other pallet: Calling a contract will transfer balance from the caller to the contract or vise versa depending on whether the call increased or decreased the storage usage. The balance is reserved in the contract's account so it cannot be used or moved away by the contract. The `deposit` made by the caller is calculated like this:
```
deposit = storage_used_before - storage_used_after
```

Note that `deposit` can be negative which constitutes a refund from the contract the caller who is removing this storage. This is not necessarily the original depositor.

One major criticism of this approach when compared with the contract based rent is that it makes contract providers inflexible with regard to their financing model because the storage is always payed by the caller. In theory, the old system allowed the contract authors to come up with their own financing model by for example pumping their own money into the contract to keep it afloat. In practice however, there are still gas costs that need to be payed by the caller so it wasn't never enough to allow fee less usage of contracts while making contract (language) development **much** harder.

I argue that the financing model of a contract should not live in the contract itself but should be provided by other means. This allows contract authors to concentrate on the business logic. Companies could provide proxy contracts to customers that are restricted in what they can do.

# Implementation Details

This is a rather simple system from the distance but there are challenges to solve in some areas.

## Contract Termination

A contract can decide to remove itself by calling the `seal_terminate` host function. As of right now a contract can call this function at any time in order to remove itself and all its associated storage.

This function will continue to work mostly like it did. It will do the following when used by a contract:

- The free balance is transferred to the specified `beneficiary` (this is unchanged)
- The storage deposits that are stored as reserved balance in the contract are transferred to the **caller**.

**tl;dr: `seal_terminate` will work fine. Free balance sent to `beneficiary`, storage deposits to the caller.**

## Code sharing

There are two kinds of storage whose size is controlled by users: 
- The key value storage of a contract that can only be accessed by the contract it belongs to. This storage is created during contract execution.
- The contract code which is a wasm blob that is uploaded to chain using the `instantiate_with_code` extrinsic and can be shared between different contracts: A contract can be instantiated without uploading any new code to the chain. Instead, it can reference an existing code by hash.

The latter is what this section is about. The code can be shared between different contract instantiations. Questions arise around the removal of those contracts. We clearly want to allow the removal of those contracts for uploaders to regain their deposit. There are two distinct challenges with regard to code sharing:

### Code blobs cannot be removed due to active users

We cannot remove code blobs which have contracts associated with them for obvious reasons. To enforce this invariant we have a reference count associated with each code blob. However a problem arises when 3rd parties start using a code blob uploaded by someone else. The uploader cannot delete the code to regain the deposit because of contracts it doesn't control.

The solution to this problem would be for contract authors to deny contract creation by entities they don't control. This can be done easily in the constructor of the contract. IMHO this is an elegant solution because it does not require baking in any logic into `pallet_contracts`. The drawback with this approach is that the default behavior would be to allow instantiation by anyone. However, that could be easily solved by contract languages that force authors to make an explicit decision in constructors. This does require any changes whatsoever to the `pallet_contracts`.

**tl;dr: We do nothing and tell contract authors to be wary about who they allow to instantiate their code blob. Also we always refund the deposit to the original uploader and not to the remover (this is different from contract storage).**

### Race between upload and instantiation

Right now there is no way to "just upload" a code blob. You need to call `instantiate_with_code` which instantiates the first contract right away. The code blob is automatically deleted when the last contract that uses it is removed. This is an elegant solution because it does not require a separate extrinsic to remove an orphaned code blob.

However, there are requests for adding an extrinsic that can upload a code blob without an associated contract (https://github.com/patractlabs/redspot/issues/136). That causes a race: Someone could delete the code hash in between the upload and instantiation.

To resolve this situation we only allow the original uploader of  a code blob to remove it. The uploader has an incentive to remove it to get the deposit back. This means that code is no longer removed automatically when the refcount drops to zero but require the submission of a `remove_code` transaction that checks whether the refcount is zero and that the submitter is equal to the original uploader.

Whenever some account uploads a code some balance gets reserved at their account (depending on the code size). This reserve is released when they delete the code. This is different from balance that is used for contract storage which is **transferred** to the contract and reserved there. This is because this balance changes hands when someone else removed the storage which is not possible for code blobs where the remover must be the original uploader.

**tl;dr: We bring pack `upload_code` and `remove_code`. The latter can only be called by the original uploader.**

## User experience

With this change a user would pay for two distinct things when calling a contract:

- Gas (weight / execution time)
- Storage

The gas is is taken from the caller as transaction fee exactly like with any other transaction. The caller can limit the amount of gas that can be used by a call using a `gas_limit`.

For the storage a deposit is made to the contract and reserved there. Both values can be estimated by pre-running the call as RPC before submitting it as a transaction. It is called an estimation because between pre-running and transaction submission the state of the chain can change and with it the behavior of the call (this is why we have `gas_limit`).

In order to allow the same for storage we add a `storage_limit` field which limits how much **balance** is allowed to be deposited as part of the call.

**tl;dr We add a `storage_limit` to the call arguments and a `storage_deposit` to the call RPC result. The latter one should be used by UIs together with `gas_used` to give the user a cost estimation. Note that `deposit_used` can be negative (refund).**

# Alternative Solutions

## State Expiry

One noteable solution to unbounded storage growth is something called **state expiry** [1](https://notes.ethereum.org/@vbuterin/verkle_and_state_expiry_proposal) [2](https://notes.ethereum.org/@vbuterin/state_expiry_eip) [3](https://hackmd.io/@vbuterin/state_expiry_paths) [4](https://ethereum-magicians.org/t/weak-statelessness-and-or-state-expiry-coming-soon/5453). The tl;dr is that full nodes  / collators only need to hold state for a fixed period of time. After that the caller / transactor is required to provide a witness for transactions that touch older states.

This is the solution which Ethereum plans to use for solving their storage growth issues. However, state expiry merely mitigates the long term consequences of wasteful behavior without discouraging it. Whether state expiry will work out is not a clear cut and might further away than we like because it requires changes to the substrate storage layer. The proposal here can be implemented relatively quickly and can be removed if it turns out too be to big of a hurdle. I am not so sure that there will be acceptance to add a deposit system into an active contract ecosystem when state expiry won't suffice. On the other hand, state expiry can be added on top at a later time without breaking existing contracts.  I speculate that this is the main reason why Ethereum opts for such a solution rather than incentivizing good behavior from contracts.

### PR DESCRIPTION
**Pull request by athei**
Fixes #9807

Please refer to the linked issue for information on this PR. I updated it recently to reflect the current state of things.

### Client Update Required

This adds new RPCs and alters existing ones with new arguments. A client update is necessary for RPC nodes.

### Storage

This PR includes changes to storage. A storage migration was added in this PR that will be executed automatically.

### UI

@jacogr @statictype 

This PR adds the new parameter `storage_deposit_limit` to existing dispatchables and RPCs. The affected dispatchables are `call`, `instantiate`, `instantiate_with_code`. Affected RPCs are `call` and `instantiate`.

This PR adds two new dispatchables `upload_code` and `remove_code` and the new RPC `upload_code`. The latter can be used to determine the `storage_deposit` required to upload some code. Those new functions should be supported eventually but do not break the UI if not added.

### Renamed `endowment` to `value`

This PR makes `0` a valid value for the `endowment` when instantiating the contract. It is no longer needed to keep the contract alive. Those costs will be deducted automatically and be limited by the `storage_deposit_limit` which can be automatically estimated through RPC. **This is why we renamed `endowment` to `value` in order to make `instantiate` consistent with `call`.**

### CODE REVIEW

File: frame/contracts/rpc/src/lib.rs [hunk 1]
```diff
@@ -89,11 +92,22 @@ pub struct InstantiateRequest<AccountId, Hash> {
...
 }
 
+/// A struct that encodes RPC parameters required for a call to upload a new code.
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct CodeUploadRequest<AccountId> {
+	origin: AccountId,
+	code: Bytes,
+	storage_limit: Option<NumberOrHex>,
```
Comment (cmichi) [inline frame/contracts/rpc/src/lib.rs:108]: I wonder if we should name this `deposit_limit` instead. It's denoted as `Balance` and afaiu limits how much deposit can be charged from the caller. To me, the term `storage_limit` indicates some limit in terms of byte size.

**Review by cmichi — APPROVED** [review]
I'm lacking a depper understanding of the codebase, but on a high-level I don't see any obvious issues. We'll need a follow-up for ink! to remove the notion of a subsistence threshold.

File: frame/contracts/rpc/src/lib.rs (hunk 1 above)
Comment (athei) [reply frame/contracts/rpc/src/lib.rs:108]: But `deposit_limit` you need to ask "deposit for what?". And using `storage_deposit_limit` is too long. Its quite the dilemma.

Comment (cmichi) [reply frame/contracts/rpc/src/lib.rs:108]: Why is it too long?

Comment (athei) [reply frame/contracts/rpc/src/lib.rs]: Fair enough. I renamed it.

File: frame/contracts/fixtures/drain.wat [hunk 2]
```diff
@@ -33,7 +33,9 @@
 			)
 		)
 
-		;; Self-destruct by sending full balance to the 0 address.
+		;; Try to self-destruct by sending full balance to the 0 address.
+		;; All the free balance will be send away but the contract's which is a valid
+		;; thing to do because the storage deposits will keep the account alive.
```
Comment (cmichi) [inline frame/contracts/fixtures/drain.wat:38]:
```suggestion
		;; All the *free* balance will be send away, which is a valid thing to do
		;; because the storage deposits will keep the account alive.
```

I had to read this a couple times to get it.

File: frame/contracts/src/wasm/prepare.rs [hunk 3]
```diff
@@ -395,20 +398,35 @@ fn check_and_instrument<C: ImportSatisfyCheck, T: Config>(
...
 		code,
-		original_code_len: original_code.len() as u32,
-		refcount: 1,
 		code_hash: T::Hashing::hash(&original_code),
 		original_code: Some(original_code),
-	})
+		owner_info: None,
+	};
+
+	// We need to add the sizes of the `#[codec(skip)]` fields which are stored in different
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/prepare.rs:416]: The comment says skipped items which are stored somewhere else. `code_hash` is skipped but stored nowhere. This is why it isn't added. If you find the text misleading please add a suggestion (I think its fine) :)

File: frame/contracts/src/wasm/mod.rs [hunk 4]
```diff
@@ -31,10 +31,13 @@ use crate::{
...
 	gas::GasMeter,
 	wasm::env_def::FunctionImplProvider,
-	CodeHash, Config, Schedule,
+	AccountIdOf, BalanceOf, CodeHash, CodeStorage, Config, Schedule,
 };
 use codec::{Decode, Encode};
-use frame_support::dispatch::DispatchError;
+use frame_support::{
+	dispatch::{DispatchError, DispatchResult},
+	pallet_prelude::MaxEncodedLen,
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/mod.rs]: Yeah I think it was moved to `parity-scale-codec` only recently.

File: frame/contracts/rpc/src/lib.rs [hunk 5]
```diff
@@ -121,7 +140,20 @@ pub trait ContractsApi<BlockHash, BlockNumber, AccountId, Balance, Hash> {
...
 		at: Option<BlockHash>,
-	) -> Result<ContractInstantiateResult<AccountId>>;
+	) -> Result<ContractInstantiateResult<AccountId, Balance>>;
+
+	/// Upload new code without instantiating a contract from it.
+	///
+	/// This upload is performed locally without submitting any transactions. Thus executing this
+	/// won't change any state.
+	///
+	/// This method is useful for calling getter-like methods on contracts.
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/rpc/src/lib.rs]: Yes!

File: frame/contracts/src/exec.rs [hunk 6]
```diff
@@ -194,6 +188,9 @@ pub trait Ext: sealing::Sealed {
 	/// Get a mutable reference to the nested gas meter.
 	fn gas_meter(&mut self) -> &mut GasMeter<Self::T>;
 
+	/// Get a mutable reference to the nested storage meter.
+	fn storage_meter(&mut self) -> &mut storage::meter::NestedMeter<Self::T>;
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/exec.rs]: A refactoring artifact I would assume. Removed it.

File: frame/contracts/src/wasm/code_cache.rs [hunk 7]
```diff
@@ -28,177 +28,170 @@
...
 ) -> Result<PrefabWasmModule<T>, DispatchError>
 where
 	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
 {
-	// The reinstrument case coincides with the cases where we need to charge extra
-	// based upon the code size: On-chain execution.
-	if let Some((_, gas_meter)) = &mut reinstrument {
-		gas_meter.charge(CodeToken::Load(estimate_code_size::<T>(&code_hash)?))?;
-	}
+	gas_meter.charge(CodeToken::Load(estimate_code_size::<T, CodeStorage<T>, _>(&code_hash)?))?;
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/code_cache.rs:152]: The function this code is in is **not** called on `upload_code`. It is only called when an existing code should be loaded from storage for execution. So only for `call` and `instantiate`. So without understanding where this question is coming from the answer is **no**.

File: frame/contracts/src/wasm/code_cache.rs [hunk 8]
```diff
@@ -28,177 +28,170 @@
...
+	let code_hash = sp_std::mem::take(&mut module.code_hash);
 	<CodeStorage<T>>::mutate(&code_hash, |existing| match existing {
-		Some(module) => increment_64(&mut module.refcount),
+		Some(existing) => {
+			// We instrument any uploaded contract anyways. We might as well store it to save
+			// a potential re-instrumentation later.
+			existing.code = module.code;
+			existing.instruction_weights_version = module.instruction_weights_version;
+			// When the code was merely uploaded but not instantiated we can skip this.
+			if instantiated {
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/code_cache.rs:63]: Yes it gives me a lot of headaches. But on the other hand if we remove it now and we need to bring it back later that would be a real PITA.

File: frame/contracts/src/wasm/code_cache.rs [hunk 9]
```diff
@@ -28,177 +28,170 @@
...
-		let original_code = <PristineCode<T>>::get(&prefab_module.code_hash)
-			.ok_or_else(|| Error::<T>::CodeNotFound)?;
-		prefab_module.code = prepare::reinstrument_contract::<T>(original_code, schedule)?;
-		prefab_module.instruction_weights_version = schedule.instruction_weights.version;
-		<CodeStorage<T>>::insert(&prefab_module.code_hash, &*prefab_module);
-		Ok(())
+	if prefab_module.instruction_weights_version < schedule.instruction_weights.version {
+		// The instruction weights have changed.
+		// We need to re-instrument the code with the new instruction weights.
+		gas_meter.charge(CodeToken::Instrument(estimate_code_size::<T, PristineCode<T>, _>(
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/code_cache.rs:161]: Yeah. If you called `upload_called` the code is upgraded to the newest `instruction_weights_version` and this charge is skipped.

File: frame/contracts/common/src/lib.rs [hunk 10]
```diff
@@ -149,3 +295,27 @@ mod as_string {
...
+#[cfg(feature = "std")]
+mod as_hex {
+	use super::*;
+	use serde::{de::Error as _, Deserializer, Serializer};
+
+	pub fn serialize<S, Balance>(balance: &Balance, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: Serializer,
+		NumberOrHex: From<Balance>,
+		Balance: Copy,
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/common/src/lib.rs]: I approve of your taste. Made the trait bounds beautiful.

File: frame/contracts/src/storage/meter.rs [hunk 11]
```diff
@@ -0,0 +1,724 @@
...
+		// We do not increase `own_deposit` because this will be charged later when the contract
+		// execution does conclude.
+		let total_deposit = self.total_deposit.saturating_add(&deposit);
+		if let Deposit::Charge(amount) = &total_deposit {
+			if amount > &self.limit {
+				return Err(<Error<T>>::StorageDepositLimitExhausted.into())
+			}
+		}
+		if let Deposit::Charge(amount) = &deposit {
+			info.storage_deposit = info.storage_deposit.saturating_add(*amount);
```
Comment (ascjones) [inline frame/contracts/src/storage/meter.rs:321]: Since this would always be true, could use `StorageDeposit::charge_or_zero` to make it a one liner

**Review by ascjones — COMMENTED** [review]
Just more superficial comments, nearly done

File: frame/contracts/src/wasm/code_cache.rs [hunk 12]
```diff
@@ -28,177 +28,170 @@
...
-		<PristineCode<T>>::insert(&code_hash, code);
-	}
+	let code_hash = sp_std::mem::take(&mut module.code_hash);
 	<CodeStorage<T>>::mutate(&code_hash, |existing| match existing {
-		Some(module) => increment_64(&mut module.refcount),
+		Some(existing) => {
+			// We instrument any uploaded contract anyways. We might as well store it to save
+			// a potential re-instrumentation later.
+			existing.code = module.code;
+			existing.instruction_weights_version = module.instruction_weights_version;
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/code_cache.rs:61]: Documented it.

File: frame/contracts/src/tests.rs [hunk 13]
```diff
@@ -1867,10 +2157,653 @@ fn ecdsa_recover() {
...
+			wasm,
+			vec![],
+			vec![],
+		));
+
+		// Drop previous events
+		initialize_block(2);
+
+		assert_noop!(
+			Contracts::remove_code(Origin::signed(BOB), code_hash),
```
Comment (ascjones) [inline frame/contracts/src/tests.rs:2363]:
```suggestion
			Contracts::remove_code(Origin::signed(ALICE), code_hash),
```

As per previous test, only origin can remove the code.

File: frame/contracts/src/tests.rs [hunk 14]
```diff
@@ -1867,10 +2157,653 @@ fn ecdsa_recover() {
...
+					}),
+					topics: vec![],
+				},
+			]
+		);
+	});
+}
+
+#[test]
+fn call_after_killed_accout_needs_funding() {
```
Comment (ascjones) [inline frame/contracts/src/tests.rs:2704]:
```suggestion
fn call_after_killed_account_needs_funding() {
```

**Review by ascjones — APPROVED** [review]
LGTM

File: frame/contracts/src/storage/meter.rs [hunk 15]
```diff
@@ -0,0 +1,724 @@
...
+		Ok(Self { limit, ..Default::default() })
+	}
+
+	/// The total amount of deposit that should change hands as result of the execution
+	/// that this meter was passed into.
+	///
+	/// This drops the root meter in order to make sure it is only called when the whole
+	/// execution did finish.
+	pub fn into_deposit(self) -> DepositOf<T> {
+		// Clone is necessary because of the drop implementation.
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/storage/meter.rs]: Refactoring artifact. There was a custom `Drop` implementation.

[the remark this answers is not in the document]
Comment (ascjones) [reply frame/contracts/src/storage/meter.rs]: Does the comment above this method also need updating "This drops the root meter in order to make sur..."

**Comment by athei** [comment]
bot merge
